File: src/PcIfy.Client/Services/FolderPrefsService.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using PcIfy.Client.Models;
using PcIfy.Client.Services.Interfaces;

namespace PcIfy.Client.Services;

public class FolderPrefsService : IFolderPrefsService
{
    private readonly string _baseDir = Path.Combine(FileSystem.AppDataDirectory, "folderprefs");

    public FolderPrefsService() => Directory.CreateDirectory(_baseDir);

    public FolderPrefs GetPrefs(string folderPath)
    {
        var file = GetFilePath(folderPath);
        if (!File.Exists(file)) return new FolderPrefs();
        try { return JsonSerializer.Deserialize<FolderPrefs>(File.ReadAllText(file)) ?? new FolderPrefs(); }
        catch { return new FolderPrefs(); }
    }

    public void SavePrefs(string folderPath, FolderPrefs prefs) =>
        File.WriteAllText(GetFilePath(folderPath), JsonSerializer.Serialize(prefs));

    private string GetFilePath(string folderPath)
    {
        var hash = Convert.ToHexString(SHA256.HashData(Encoding.UTF8.GetBytes(folderPath)));
        return Path.Combine(_baseDir, hash + ".json");
    }
}

File: src/PcIfy.Client/lib/services/folder_prefs_service.dart
import 'dart:convert';
import 'dart:io';

import 'package:crypto/crypto.dart';
import 'package:path_provider/path_provider.dart';

import '../core/models/folder_prefs.dart';

class FolderPrefsService {
  Directory? _dir;

  Future<Directory> _getDir() async {
    if (_dir != null) return _dir!;
    final base = await getApplicationDocumentsDirectory();
    final dir = Directory('${base.path}/folderprefs');
    if (!await dir.exists()) await dir.create(recursive: true);
    _dir = dir;
    return dir;
  }

  String _key(String folderPath) {
    final bytes = utf8.encode(folderPath);
    return sha256.convert(bytes).toString();
  }

  Future<FolderPrefs> getPrefs(String folderPath) async {
    final dir = await _getDir();
    final file = File('${dir.path}/${_key(folderPath)}.json');
    if (!await file.exists()) return const FolderPrefs();
    try {
      final json = jsonDecode(await file.readAsString()) as Map<String, dynamic>;
      return FolderPrefs.fromJson(json);
    } catch (_) {
      return const FolderPrefs();
    }
  }

  Future<void> savePrefs(String folderPath, FolderPrefs prefs) async {
    final dir = await _getDir();
    final file = File('${dir.path}/${_key(folderPath)}.json');
    await file.writeAsString(jsonEncode(prefs.toJson()));
  }

  Future<void> savePrefsForHash(String hash, FolderPrefs prefs) async {
    final dir = await _getDir();
    final file = File('${dir.path}/$hash.json');
    await file.writeAsString(jsonEncode(prefs.toJson()));
  }

  Future<Map<String, FolderPrefs>> getAllPrefs() async {
    final dir = await _getDir();
    final result = <String, FolderPrefs>{};
    await for (final entity in dir.list()) {
      if (entity is! File || !entity.path.endsWith('.json')) continue;
      final hash = entity.uri.pathSegments.last.replaceAll('.json', '');
      try {
        final json =
            jsonDecode(await entity.readAsString()) as Map<String, dynamic>;
        result[hash] = FolderPrefs.fromJson(json);
      } catch (_) {}
    }
    return result;
  }
}
